File: 0763-partition-labels/0763-partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        intervals = {}
        for i, c in enumerate(s):
            if c not in intervals:
                intervals[c] = [i,i]
            else:
                intervals[c][1] = i
        intervals = list(intervals.values())
        intervals.sort()
        start, end = intervals[0][0], intervals[0][1]
        res = []
        for i in range(1, len(intervals)):
            if intervals[i][0] > end: # if interval is disjoint
                res.append(end - start + 1)
                start, end = intervals[i][0], intervals[i][1]
            else:
                end = max(end, intervals[i][1])
        res.append(end - start + 1)
        return res
        

File: 0763-partition-labels/0763-partition-labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        int n=s.length();
        unordered_map<char,int> mp;
        for(int i=0;i<n;i++){
            mp[s[i]]=i;
        }
        vector<int> v;
        int l=0,r=mp[s[l]],prev=0;
        while(l<n && r<n){
            while(l<r){
                if(mp[s[l]]>r){
                    r=mp[s[l]];
                }
                l++;
            }
            if(l==r){
                v.push_back(r-prev+1);
                prev=l+1;
                l++;
                r=mp[s[l]];
            }
        }
        return v;
    }
};